### ISSUE TITLE
aws-cdk: sam local error on generated template

### ISSUE THREAD

**Issue by myin142** [aws/aws-sam-cli#2287]
When I try to run `sam local start-api` with the generated template from aws-cdk, I get the following error when I make an OPTIONS request to the `/time` endpoint.

Error:

```
Exception on /time [OPTIONS]
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/lib/python3.8/site-packages/samcli/local/apigw/local_apigw_service.py", line 203, in _request_handler
    self.lambda_runner.invoke(route.function_name, event, stdout=stdout_stream_writer, stderr=self.stderr)
  File "/usr/lib/python3.8/site-packages/samcli/commands/local/lib/local_lambda.py", line 84, in invoke
    function = self.provider.get(function_name)
  File "/usr/lib/python3.8/site-packages/samcli/lib/providers/sam_function_provider.py", line 65, in get
    raise ValueError("Function name is required")
ValueError: Function name is required
```

The template file:

```
Resources:
  WorkTrackerBC4B47B0:
    Type: AWS::DynamoDB::Table
    Properties:
      KeySchema:
        - AttributeName: user
          KeyType: HASH
        - AttributeName: timestamp
          KeyType: RANGE
      AttributeDefinitions:
        - AttributeName: user
          AttributeType: S
        - AttributeName: timestamp
          AttributeType: S
      ProvisionedThroughput:
        ReadCapacityUnits: 5
        WriteCapacityUnits: 5
      TableName: WorkTracker
    UpdateReplacePolicy: Retain
    DeletionPolicy: Retain
    Metadata:
      aws:cdk:path: work-tracker/WorkTracker/Resource
  WorkTrackerApiA3820895:
    Type: AWS::ApiGateway::RestApi
    Properties:
      Name: WorkTrackerApi
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Resource
  WorkTrackerApiCloudWatchRole46BF1BDC:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Action: sts:AssumeRole
            Effect: Allow
            Principal:
              Service: apigateway.amazonaws.com
        Version: "2012-10-17"
      ManagedPolicyArns:
        - Fn::Join:
            - ""
            - - "arn:"
              - Ref: AWS::Partition
              - :iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/CloudWatchRole/Resource
  WorkTrackerApiAccountB5820D17:
    Type: AWS::ApiGateway::Account
    Properties:
      CloudWatchRoleArn:
        Fn::GetAtt:
          - WorkTrackerApiCloudWatchRole46BF1BDC
          - Arn
    DependsOn:
      - WorkTrackerApiA3820895
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Account
  WorkTrackerApiDeployment667E8F39a4dede785b4b39d382183529b3eb56f9:
    Type: AWS::ApiGateway::Deployment
    Properties:
      RestApiId:
        Ref: WorkTrackerApiA3820895
      Description: Automatically created by the RestApi construct
    DependsOn:
      - WorkTrackerApiOPTIONS4CF25EFB
      - WorkTrackerApitimeGETAC751B15
      - WorkTrackerApitimeOPTIONS121CA0A4
      - WorkTrackerApitimePOSTB776371B
      - WorkTrackerApitimeE76CB408
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Deployment/Resource
  WorkTrackerApiDeploymentStageprodE0E1DD82:
    Type: AWS::ApiGateway::Stage
    Properties:
      RestApiId:
        Ref: WorkTrackerApiA3820895
      DeploymentId:
        Ref: WorkTrackerApiDeployment667E8F39a4dede785b4b39d382183529b3eb56f9
      MethodSettings:
        - HttpMethod: "*"
          ResourcePath: /*
          ThrottlingBurstLimit: 10
          ThrottlingRateLimit: 10
      StageName: prod
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/DeploymentStage.prod/Resource
  WorkTrackerApiOPTIONS4CF25EFB:
    Type: AWS::ApiGateway::Method
    Properties:
      HttpMethod: OPTIONS
      ResourceId:
        Fn::GetAtt:
          - WorkTrackerApiA3820895
          - RootResourceId
      RestApiId:
        Ref: WorkTrackerApiA3820895
      AuthorizationType: NONE
      Integration:
        IntegrationResponses:
          - ResponseParameters:
              method.response.header.Access-Control-Allow-Headers: "'Authorization'"
              method.response.header.Access-Control-Allow-Origin: "'*'"
              method.response.header.Access-Control-Allow-Methods: "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'"
              method.response.header.Access-Control-Allow-Credentials: "'true'"
            StatusCode: "204"
        RequestTemplates:
          application/json: "{ statusCode: 200 }"
        Type: MOCK
      MethodResponses:
        - ResponseParameters:
            method.response.header.Access-Control-Allow-Headers: true
            method.response.header.Access-Control-Allow-Origin: true
            method.response.header.Access-Control-Allow-Methods: true
            method.response.header.Access-Control-Allow-Credentials: true
          StatusCode: "204"
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Default/OPTIONS/Resource
  WorkTrackerApitimeE76CB408:
    Type: AWS::ApiGateway::Resource
    Properties:
      ParentId:
        Fn::GetAtt:
          - WorkTrackerApiA3820895
          - RootResourceId
      PathPart: time
      RestApiId:
        Ref: WorkTrackerApiA3820895
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Default/time/Resource
  WorkTrackerApitimeOPTIONS121CA0A4:
    Type: AWS::ApiGateway::Method
    Properties:
      HttpMethod: OPTIONS
      ResourceId:
        Ref: WorkTrackerApitimeE76CB408
      RestApiId:
        Ref: WorkTrackerApiA3820895
      AuthorizationType: NONE
      Integration:
        IntegrationResponses:
          - ResponseParameters:
              method.response.header.Access-Control-Allow-Headers: "'Authorization'"
              method.response.header.Access-Control-Allow-Origin: "'*'"
              method.response.header.Access-Control-Allow-Methods: "'OPTIONS,GET,PUT,POST,DELETE,PATCH,HEAD'"
              method.response.header.Access-Control-Allow-Credentials: "'true'"
            StatusCode: "204"
        RequestTemplates:
          application/json: "{ statusCode: 200 }"
        Type: MOCK
      MethodResponses:
        - ResponseParameters:
            method.response.header.Access-Control-Allow-Headers: true
            method.response.header.Access-Control-Allow-Origin: true
            method.response.header.Access-Control-Allow-Methods: true
            method.response.header.Access-Control-Allow-Credentials: true
          StatusCode: "204"
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Default/time/OPTIONS/Resource
  WorkTrackerApitimeGETApiPermissionworktrackerWorkTrackerApi0D6A680AGETtime5350A293:
    Type: AWS::Lambda::Permission
    Properties:
      Action: lambda:InvokeFunction
      FunctionName:
        Fn::GetAtt:
          - getTimesOfDayC45A06A5
          - Arn
      Principal: apigateway.amazonaws.com
      SourceArn:
        Fn::Join:
          - ""
          - - "arn:"
            - Ref: AWS::Partition
            - ":execute-api:"
            - Ref: AWS::Region
            - ":"
            - Ref: AWS::AccountId
            - ":"
            - Ref: WorkTrackerApiA3820895
            - /
            - Ref: WorkTrackerApiDeploymentStageprodE0E1DD82
            - /GET/time
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Default/time/GET/ApiPermission.worktrackerWorkTrackerApi0D6A680A.GET..time
  WorkTrackerApitimeGETApiPermissionTestworktrackerWorkTrackerApi0D6A680AGETtime05C7716C:
    Type: AWS::Lambda::Permission
    Properties:
      Action: lambda:InvokeFunction
      FunctionName:
        Fn::GetAtt:
          - getTimesOfDayC45A06A5
          - Arn
      Principal: apigateway.amazonaws.com
      SourceArn:
        Fn::Join:
          - ""
          - - "arn:"
            - Ref: AWS::Partition
            - ":execute-api:"
            - Ref: AWS::Region
            - ":"
            - Ref: AWS::AccountId
            - ":"
            - Ref: WorkTrackerApiA3820895
            - /test-invoke-stage/GET/time
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Default/time/GET/ApiPermission.Test.worktrackerWorkTrackerApi0D6A680A.GET..time
  WorkTrackerApitimeGETAC751B15:
    Type: AWS::ApiGateway::Method
    Properties:
      HttpMethod: GET
      ResourceId:
        Ref: WorkTrackerApitimeE76CB408
      RestApiId:
        Ref: WorkTrackerApiA3820895
      AuthorizationType: COGNITO_USER_POOLS
      AuthorizerId:
        Ref: CognitoAuthorizer
      Integration:
        IntegrationHttpMethod: POST
        Type: AWS_PROXY
        Uri:
          Fn::Join:
            - ""
            - - "arn:"
              - Ref: AWS::Partition
              - ":apigateway:"
              - Ref: AWS::Region
              - :lambda:path/2015-03-31/functions/
              - Fn::GetAtt:
                  - getTimesOfDayC45A06A5
                  - Arn
              - /invocations
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Default/time/GET/Resource
  WorkTrackerApitimePOSTApiPermissionworktrackerWorkTrackerApi0D6A680APOSTtime9B216B47:
    Type: AWS::Lambda::Permission
    Properties:
      Action: lambda:InvokeFunction
      FunctionName:
        Fn::GetAtt:
          - getTimesOfDayC45A06A5
          - Arn
      Principal: apigateway.amazonaws.com
      SourceArn:
        Fn::Join:
          - ""
          - - "arn:"
            - Ref: AWS::Partition
            - ":execute-api:"
            - Ref: AWS::Region
            - ":"
            - Ref: AWS::AccountId
            - ":"
            - Ref: WorkTrackerApiA3820895
            - /
            - Ref: WorkTrackerApiDeploymentStageprodE0E1DD82
            - /POST/time
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Default/time/POST/ApiPermission.worktrackerWorkTrackerApi0D6A680A.POST..time
  WorkTrackerApitimePOSTApiPermissionTestworktrackerWorkTrackerApi0D6A680APOSTtimeA2323B78:
    Type: AWS::Lambda::Permission
    Properties:
      Action: lambda:InvokeFunction
      FunctionName:
        Fn::GetAtt:
          - getTimesOfDayC45A06A5
          - Arn
      Principal: apigateway.amazonaws.com
      SourceArn:
        Fn::Join:
          - ""
          - - "arn:"
            - Ref: AWS::Partition
            - ":execute-api:"
            - Ref: AWS::Region
            - ":"
            - Ref: AWS::AccountId
            - ":"
            - Ref: WorkTrackerApiA3820895
            - /test-invoke-stage/POST/time
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Default/time/POST/ApiPermission.Test.worktrackerWorkTrackerApi0D6A680A.POST..time
  WorkTrackerApitimePOSTB776371B:
    Type: AWS::ApiGateway::Method
    Properties:
      HttpMethod: POST
      ResourceId:
        Ref: WorkTrackerApitimeE76CB408
      RestApiId:
        Ref: WorkTrackerApiA3820895
      AuthorizationType: COGNITO_USER_POOLS
      AuthorizerId:
        Ref: CognitoAuthorizer
      Integration:
        IntegrationHttpMethod: POST
        Type: AWS_PROXY
        Uri:
          Fn::Join:
            - ""
            - - "arn:"
              - Ref: AWS::Partition
              - ":apigateway:"
              - Ref: AWS::Region
              - :lambda:path/2015-03-31/functions/
              - Fn::GetAtt:
                  - getTimesOfDayC45A06A5
                  - Arn
              - /invocations
    Metadata:
      aws:cdk:path: work-tracker/WorkTrackerApi/Default/time/POST/Resource
  getTimesOfDayServiceRole7C5D69EC:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Action: sts:AssumeRole
            Effect: Allow
            Principal:
              Service: lambda.amazonaws.com
        Version: "2012-10-17"
      ManagedPolicyArns:
        - Fn::Join:
            - ""
            - - "arn:"
              - Ref: AWS::Partition
              - :iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
    Metadata:
      aws:cdk:path: work-tracker/getTimesOfDay/ServiceRole/Resource
  getTimesOfDayServiceRoleDefaultPolicy371E6EE6:
    Type: AWS::IAM::Policy
    Properties:
      PolicyDocument:
        Statement:
          - Action:
              - dynamodb:BatchGetItem
              - dynamodb:GetRecords
              - dynamodb:GetShardIterator
              - dynamodb:Query
              - dynamodb:GetItem
              - dynamodb:Scan
            Effect: Allow
            Resource:
              - Fn::GetAtt:
                  - WorkTrackerBC4B47B0
                  - Arn
              - Ref: AWS::NoValue
        Version: "2012-10-17"
      PolicyName: getTimesOfDayServiceRoleDefaultPolicy371E6EE6
      Roles:
        - Ref: getTimesOfDayServiceRole7C5D69EC
    Metadata:
      aws:cdk:path: work-tracker/getTimesOfDay/ServiceRole/DefaultPolicy/Resource
  getTimesOfDayC45A06A5:
    Type: AWS::Lambda::Function
    Properties:
      Code:
        S3Bucket:
          Ref: AssetParametersf4569dbf4276b2d1eec276690859fd4376007e4aaf42b4f7330451c1822781baS3BucketB50D3E5E
        S3Key:
          Fn::Join:
            - ""
            - - Fn::Select:
                  - 0
                  - Fn::Split:
                      - "||"
                      - Ref: AssetParametersf4569dbf4276b2d1eec276690859fd4376007e4aaf42b4f7330451c1822781baS3VersionKey9E270B2B
              - Fn::Select:
                  - 1
                  - Fn::Split:
                      - "||"
                      - Ref: AssetParametersf4569dbf4276b2d1eec276690859fd4376007e4aaf42b4f7330451c1822781baS3VersionKey9E270B2B
      Handler: daily-time.handler
      Role:
        Fn::GetAtt:
          - getTimesOfDayServiceRole7C5D69EC
          - Arn
      Runtime: nodejs12.x
    DependsOn:
      - getTimesOfDayServiceRoleDefaultPolicy371E6EE6
      - getTimesOfDayServiceRole7C5D69EC
    Metadata:
      aws:cdk:path: work-tracker/getTimesOfDay/Resource
      aws:asset:path: /home/sakkaku/Documents/workspaces/myin-work/dist/libs/daily-time
      aws:asset:property: Code
  CognitoAuthorizer:
    Type: AWS::ApiGateway::Authorizer
    Properties:
      RestApiId:
        Ref: WorkTrackerApiA3820895
      Type: COGNITO_USER_POOLS
      IdentitySource: method.request.header.Authorization
      Name: MainPoolAuthorizer
      ProviderARNs: []
    Metadata:
      aws:cdk:path: work-tracker/CognitoAuthorizer
  CDKMetadata:
    Type: AWS::CDK::Metadata
    Properties:
      Modules: aws-cdk=1.65.0,@aws-cdk/assets=1.65.0,@aws-cdk/aws-apigateway=1.65.0,@aws-cdk/aws-applicationautoscaling=1.65.0,@aws-cdk/aws-autoscaling-common=1.65.0,@aws-cdk/aws-cloudwatch=1.65.0,@aws-cdk/aws-codeguruprofiler=1.65.0,@aws-cdk/aws-dynamodb=1.65.0,@aws-cdk/aws-ec2=1.65.0,@aws-cdk/aws-events=1.65.0,@aws-cdk/aws-iam=1.65.0,@aws-cdk/aws-kms=1.65.0,@aws-cdk/aws-lambda=1.65.0,@aws-cdk/aws-logs=1.65.0,@aws-cdk/aws-s3=1.65.0,@aws-cdk/aws-s3-assets=1.65.0,@aws-cdk/aws-sqs=1.65.0,@aws-cdk/aws-ssm=1.65.0,@aws-cdk/cloud-assembly-schema=1.65.0,@aws-cdk/core=1.65.0,@aws-cdk/custom-resources=1.65.0,@aws-cdk/cx-api=1.65.0,@aws-cdk/region-info=1.65.0,jsii-runtime=node.js/v14.10.1
    Metadata:
      aws:cdk:path: work-tracker/CDKMetadata/Default
    Condition: CDKMetadataAvailable
Outputs:
  WorkTrackerApiEndpoint24C3B084:
    Value:
      Fn::Join:
        - ""
        - - https://
          - Ref: WorkTrackerApiA3820895
          - .execute-api.
          - Ref: AWS::Region
          - "."
          - Ref: AWS::URLSuffix
          - /
          - Ref: WorkTrackerApiDeploymentStageprodE0E1DD82
          - /
Parameters:
  AssetParametersf4569dbf4276b2d1eec276690859fd4376007e4aaf42b4f7330451c1822781baS3BucketB50D3E5E:
    Type: String
    Description: S3 bucket for asset "f4569dbf4276b2d1eec276690859fd4376007e4aaf42b4f7330451c1822781ba"
  AssetParametersf4569dbf4276b2d1eec276690859fd4376007e4aaf42b4f7330451c1822781baS3VersionKey9E270B2B:
    Type: String
    Description: S3 key for asset version "f4569dbf4276b2d1eec276690859fd4376007e4aaf42b4f7330451c1822781ba"
  AssetParametersf4569dbf4276b2d1eec276690859fd4376007e4aaf42b4f7330451c1822781baArtifactHashD8875F9F:
    Type: String
    Description: Artifact hash for asset "f4569dbf4276b2d1eec276690859fd4376007e4aaf42b4f7330451c1822781ba"
Conditions:
  CDKMetadataAvailable:
    Fn::Or:
      - Fn::Or:
          - Fn::Equals:
              - Ref: AWS::Region
              - ap-east-1
          - Fn::Equals:
              - Ref: AWS::Region
              - ap-northeast-1
          - Fn::Equals:
              - Ref: AWS::Region
              - ap-northeast-2
          - Fn::Equals:
              - Ref: AWS::Region
              - ap-south-1
          - Fn::Equals:
              - Ref: AWS::Region
              - ap-southeast-1
          - Fn::Equals:
              - Ref: AWS::Region
              - ap-southeast-2
          - Fn::Equals:
              - Ref: AWS::Region
              - ca-central-1
          - Fn::Equals:
              - Ref: AWS::Region
              - cn-north-1
          - Fn::Equals:
              - Ref: AWS::Region
              - cn-northwest-1
          - Fn::Equals:
              - Ref: AWS::Region
              - eu-central-1
      - Fn::Or:
          - Fn::Equals:
              - Ref: AWS::Region
              - eu-north-1
          - Fn::Equals:
              - Ref: AWS::Region
              - eu-west-1
          - Fn::Equals:
              - Ref: AWS::Region
              - eu-west-2
          - Fn::Equals:
              - Ref: AWS::Region
              - eu-west-3
          - Fn::Equals:
              - Ref: AWS::Region
              - me-south-1
          - Fn::Equals:
              - Ref: AWS::Region
              - sa-east-1
          - Fn::Equals:
              - Ref: AWS::Region
              - us-east-1
          - Fn::Equals:
              - Ref: AWS::Region
              - us-east-2
          - Fn::Equals:
              - Ref: AWS::Region
              - us-west-1
          - Fn::Equals:
              - Ref: AWS::Region
              - us-west-2


```

**Comment by mndeveci** [comment]
Hi @myin142 ,

I tried to re-produce your issue with the following CDK project and steps;

This is my CDK stack;
```
import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as apigateway from "@aws-cdk/aws-apigateway";


export class CdkSamExampleStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const handler = new lambda.Function(this, 'MyFunction', {
      runtime: lambda.Runtime.PYTHON_3_7,
      handler: 'app.lambda_handler',
      code:    lambda.Code.fromAsset('./my_function'),
    });

    const api = new apigateway.RestApi(this, "widgets-api", {
      restApiName: "Widget Service",
      description: "This service serves widgets."
    });

    const getWidgetsIntegration = new apigateway.LambdaIntegration(handler, {
      requestTemplates: { "application/json": '{ "statusCode": "200" }' }
    });

    api.root.addMethod("GET", getWidgetsIntegration); // GET
    api.root.addMethod("OPTIONS", getWidgetsIntegration); // OPTIONS
  }
}
```

And this is the function;
```
import json

def lambda_handler(event, context):
  return {
        "statusCode": 200,
        "body": json.dumps({
            "message": "hello world",
        }),
    }
```

And when I run the following commands, I can call `GET` endpoint but `OPTIONS` is not working;
`cdk synth --no-staging > template.yaml && /usr/local/bin/sam local start-api`

I tried same example with pure SAM application with following function configuration, and I was able to call both `GET` and `OPTIONS` endpoint;
```
Resources:
  HelloWorldFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: hello_world/
      Handler: app.lambda_handler
      Runtime: python3.8
      Events:
        HelloWorld:
          Type: Api
          Properties:
            Path: /hello
            Method: get
        HelloWorldOptions:
          Type: Api
          Properties:
            Path: /hello
            Method: options
```

I am not sure if I am setting up CDK example correct, but can you give us more details about your stack? A locally reproducible example will be helpful.

Thanks!

**Comment by myin142** [comment]
@mndeveci I don't understand your example. You say OPTIONS is not working but also mention that your CDK example is correct. Is it now working or not?

Regardless, I tried to create a minimal example to reproduce the bug. It looks like the problem is the `defaultCorsPreflightOptions`.

Stack:
```
import * as cdk from '@aws-cdk/core';
import { Function, Runtime, Code } from '@aws-cdk/aws-lambda';
import { LambdaIntegration, RestApi } from '@aws-cdk/aws-apigateway';

export class AwsCdkExampleStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const api = new RestApi(this, 'Api', {
      defaultCorsPreflightOptions: {
        allowOrigins: ['*'],
      },
    });

    const lambda = new Function(this, 'Lambda', {
      runtime: Runtime.NODEJS_12_X,
      code: Code.fromAsset('resources'),
      handler: 'app.handler',
    });

    api.root.addMethod('GET', new LambdaIntegration(lambda));
  }
}

```

Lambda file:
```
exports.handler = async (ev, ctx) => {
    return {
        statusCode: 200,
        body: JSON.stringify({ success: true }),
    };
}
```

**Comment by akomiqaia** [comment]
Hi. I also have the same error when I try to run `sam local start-api`. `get` methods work as expected but `OPTIONS` method returns 502 status code and error: `Function name is required`. @myin142 why do you think it might be `defaultCorsPreflightOptions` fault and did you manage to find a workaround this one?

**Comment by myin142** [comment]
@akomiqaia Because of the minimal example that I tested it with. It works if I remove the `defaultCorsPreflightOptions` option. When I tried to add the OPTIONS like mndeveci did it, then it seemed to work. Unfortunately I did not find any workarounds for it except replacing the `defaultCorsPreflightOptions` with manual OPTIONS methods for all endpoints.

**Comment by swizzmagik** [comment]
@myin142 I am running into the same issue. Do you mind sharing an example of how you manually set the OPTIONS methods in your solution? Thanks!

**Comment by myin142** [comment]
@swizzmagik Just like mndeveci has done it
```
api.root.addMethod("GET", getWidgetsIntegration); // GET
api.root.addMethod("OPTIONS", getWidgetsIntegration); // OPTIONS
```

**Comment by swizzmagik** [comment]
> @swizzmagik Just like mndeveci has done it
> 
> ```
> api.root.addMethod("GET", getWidgetsIntegration); // GET
> api.root.addMethod("OPTIONS", getWidgetsIntegration); // OPTIONS
> ```

Thanks, that allowed my OPTIONS request to complete with 200, but I am still getting this:

Access to fetch at 'https://xxxxxx.ngrok.io' from origin 'https://yyyyyy.ngrok.io' has been blocked by CORS policy: Request header field content-type is not allowed by Access-Control-Allow-Headers in preflight response.

anything I try to do to explicitly enable those additional preflight settings leads to the same: 

ValueError: Function name is required 

error. 

I have tried adding through the `addCorsPreflightoption `on the resource as well as the `defaultCorsPreflightOptions` on the RestApi itself with the same result.

**Comment by myin142** [comment]
I am not sure if it is possible but in the example code you could try to add the CORS header yourself. So in the Lambda of `getWidgetsIntegration` you just add the header `Access-Control-Allow-Headers: *` or something like that.

**Comment by revmischa** [comment]
I'm also using CDK and getting this error:
```
[start:infra] Mounting None at http://127.0.0.1:4010/tenant [GET, POST, OPTIONS]
[start:infra] You can now browse to the above endpoints to invoke your functions. You do not need to restart/reload SAM CLI while working on your functions, changes will be reflected instantly/automatically. You only need to restart SAM CLI if you update your AWS SAM template
[start:infra] 2021-09-02 15:13:03  * Running on http://127.0.0.1:4010/ (Press CTRL+C to quit)
[start:infra] Exception on /tenant [GET]
[start:infra] Traceback (most recent call last):
[start:infra]   File "/usr/local/Cellar/aws-sam-cli-beta-cdk/202108311500/libexec/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
[start:infra]     response = self.full_dispatch_request()
[start:infra]   File "/usr/local/Cellar/aws-sam-cli-beta-cdk/202108311500/libexec/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
[start:infra]     rv = self.handle_user_exception(e)
[start:infra]   File "/usr/local/Cellar/aws-sam-cli-beta-cdk/202108311500/libexec/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
[start:infra]     reraise(exc_type, exc_value, tb)
[start:infra]   File "/usr/local/Cellar/aws-sam-cli-beta-cdk/202108311500/libexec/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
[start:infra]     raise value
[start:infra]   File "/usr/local/Cellar/aws-sam-cli-beta-cdk/202108311500/libexec/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
[start:infra]     rv = self.dispatch_request()
[start:infra]   File "/usr/local/Cellar/aws-sam-cli-beta-cdk/202108311500/libexec/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
[start:infra]     return self.view_functions[rule.endpoint](**req.view_args)
[start:infra]   File "/usr/local/Cellar/aws-sam-cli-beta-cdk/202108311500/libexec/lib/python3.8/site-packages/samcli/local/apigw/local_apigw_service.py", line 317, in _request_handler
[start:infra]     self.lambda_runner.invoke(route.function_name, event, stdout=stdout_stream_writer, stderr=self.stderr)
[start:infra]   File "/usr/local/Cellar/aws-sam-cli-beta-cdk/202108311500/libexec/lib/python3.8/site-packages/samcli/commands/local/lib/local_lambda.py", line 107, in invoke
[start:infra]     function = self.provider.get(function_identifier)
[start:infra]   File "/usr/local/Cellar/aws-sam-cli-beta-cdk/202108311500/libexec/lib/python3.8/site-packages/samcli/lib/providers/sam_function_provider.py", line 74, in get
[start:infra]     raise ValueError("Function name is required")
[start:infra] ValueError: Function name is required
```

Removing my `defaultCorsPreflightOptions` didn't help, although CORS support does seem lacking in general for local development.

My functions definitely have well-defined names which is the weird thing.

**Comment by c0pp3rt0p** [comment]
Currently SAM still doesn't support `defaultCorsPreflightOptions` and you need to manually add the `OPTIONS` method as @myin142 mentions above.  

@mildaniel PR does seem to address the issue, but the PR is languishing.

**Comment by torresxb1** [comment]
Closing as this should be addressed in the latest release, 1.47.0. 
Please let us know if you still experience issues. Thanks!
